More general equalsModProperty

tobias-rnh opened the pull request:
<!--
Thanks for submitting this pull request for KeY.
Since the project has a strict review policy, please make the
reviewer's job easier by providing the necessary information
in the text below. The comments can be deleted, but may also 
remain since they will be invisible when showing the PR.
-->

## Intended Change

<!-- Please give a brief description of what behaviour changes and 
     why it should be changed. -->
Building on top of #3386, the interface `EqualsModProperty` has now been changed so that it can be used not just with `Term`s but theoretically any class if a proper `Property` is implemented. The signature of the method has also been slightly modified so that it is able to handle equality checks that utilize additional parameters (like was the case with `equalsModRenaming` defined on `SourceElement`s that is also refactored with this PR).
~~To refactor `equalsModRenaming` of `SourceElement`, a new Interface `TreeWalker` and implementing class `JavaASTTreeWalker` have been introduced that make it possible to move through the tree of  SourceElements step by step.~~
The new general navigation structure is used to move through the tree of SourceElements.

## Plan

<!-- If this pull request is not yet finished, but has open jobs,
     please list them here. It helps others to estimate the state of this
     PR. This list can and should be adapted as the PR develops.
     Items should be crossed out when finished, not deleted.
     
     Remove the section if the PR was finished at submission time.
     The following items are examples:
-->
~~Direct tests for JavaASTTreeWalker (currently only indirectly tested through tests for equalsModProperty)~~
~~Even more doc comments~~

## Type of pull request

<!--- What types of changes does your code introduce? 
      Delete those lines that do not apply.      
-->

- Refactoring (behaviour should not change or only minimally change)
- New feature (non-breaking change which adds functionality)
- There are changes to the (Java) code

## Ensuring quality

<!--- How did you make sure that the proposed change improves the code base? 
      Delete those lines that do not apply.
      Please make an effort to delete as few lines as possible. :-)
-->
    
- I made sure that introduced/changed code is well documented (javadoc and inline comments).
- I added new test case(s) for new functionality. 
- I have tested the feature as follows: Already existing and new tests

## Additional information and contact(s)

<!-- Add further information to help the reviewer understand the request.
     Leave empty if you are sure the reviewer does not need more
     
     Who apart from yourself is involved in this pull request?
     Use @mentions to refer to them here.
     Use Co-Authored-By in your git commits if applicable. -->

### EqualsModProperty

The old `TermEqualsModProperty`
```java
public interface TermEqualsModProperty {
    boolean equalsModProperty(Object o, TermProperty property);
    int hashCodeModProperty(TermProperty property);
}
```
has now been parameterized with a type `T` so that it can accommodate other classes than just `Term`:
```java
public interface EqualsModProperty<T> {
    <V> boolean equalsModProperty(Object o, Property<T> property, V... v);
    int hashCodeModProperty(Property<T> property);
}
```
`equalsModProperty` is furthermore parameterized with a type `V` that can account for additional parameters needed for the equality check. 

### Properties
```java
public interface TermProperty {
    Boolean equalsModThisProperty(Term term1, Term term2);
    int hashCodeModThisProperty(Term term);
}
```
is now parameterized with a type `T` so that it can accommodate other classes than just `Term`:

```java
public interface Property<T> {
    <V> boolean equalsModThisProperty(T t1, T t2, V... v);
    int hashCodeModThisProperty(T t);
}
```
`equalsModThisProperty` is furthermore parameterized with a type `V` that can account for additional parameters needed for the equality check.

### Refactoring of equalsModRenaming

As `equals` and `equalsModRenaming` for SourceElements were basically doing the same thing except for a few classes (`equals` mostly called `equalsModRenaming`), the idea was that the content of `equalsModRenaming` is moved to `equals` entirely. The special cases that did not call `equalsModRenaming` in `equals` or made use of the `NameAbstractionTable` are now handled by a `JavaASTTreeWalker`.
     
<!-- DRAFT MODE: Please note that on the button to submit this pull
     request you can select between submitting a merge-ready request
     or one in draft mode (still evolving). 
     Please use the draft mode unless you think that your proposal
     should be brought onto master in the current form. -->

The contributions within this pull request are licensed under GPLv2 (only) for inclusion in KeY.

tobias-rnh commented:
This time the unit tests for windows fail but they run on my machine.
The CodeQuality checks `checkstyle_new` and `pmd` seem to also not be working now, however the error messages seem to not be related to my changes.

Drodt commented:
3 things to do before merge:
1. Fix merge conflicts.
2. Some Classes (e.g., `StatementBlock`) now have `equals` methods but no corresponding `hash` methods. Those would need to be added.
3. Add default methods to `EqualsModProperty` interface like `equalsModRemaning` that call `equalsModProperty` with the corresponding property.

tobias-rnh commented on key.core/src/main/java/de/uka/ilkd/key/ldt/FreeLDT.java:
```diff
@@ -22,7 +22,7 @@
  * additional abstract data type, e.g., binary trees, stacks, etc. in <code>.key</code> files.
  *
- * @author daniel
+ * @author Daniel Bruns
```
I have changed the name.

tobias-rnh commented:
> 3 things to do before merge:
> 
>     1. Fix merge conflicts.
Done.
> 
>     2. Some Classes (e.g., `StatementBlock`) now have `equals` methods but no corresponding `hash` methods. Those would need to be added.
Most of the time, this was a false alert. The hashCode() method is not directly overridden, but the method computeHashCode() that does the actual computation in hashCode() is overridden. For some classes, however, I have added implementations of computeHashCode().
> 
>     3. Add default methods to `EqualsModProperty` interface like `equalsModRemaning` that call `equalsModProperty` with the corresponding property.
As discussed, we do not add default methods. Most of the properties like TermLabelsProperty are only defined for terms, and there are no comparable properties defined for other instances of the generic type T.

tobias-rnh commented:
I have changed equalsModThisProperty in RenamingSourceElementProperty to use the general navigation structure instead of the TreeWalker. Because of this, I have removed the TreeWalker entirely.
This PR could now be merged after all the tests have passed.
